### ISSUE TITLE
VirgilNN: Vision-based Image Recognition Geocoder for Identifying Locations using a Neural Network

### ISSUE THREAD

**Issue by momir64** [ftn-ai-lab/sc-2025#1]
## Tim
- Luka Bursać SV22/2022 
- Momir Stanišić SV39/2022

## Asistent
- Dragan Vidaković

<br/>
<br/>

[predlog_projekta_virgilnn.pdf](https://github.com/user-attachments/files/25169736/predlog_projekta_virgilnn.pdf)

**Comment by vdragan1993** [comment]
Potrebno je da otkucate sadržaj predloga projekta u sklopu *issue*-a prateći smernice iz dokumenta okačenog na *Canvas*-u. Fokusirajte se na suštinu prilikom pisanja.

**Comment by momir64** [comment]
# VirgilNN: Vision-based Image Recognition Geocoder for Identifying Locations using a Neural Network

## Tim
- Luka Bursać SV22/2022 
- Momir Stanišić SV39/2022

## Asistent
- Dragan Vidaković

## Definicija problema

Problem geokodovanja predstavlja problem određivanja GPS koordinata na osnovu datih fotografija. Projekat predstavlja proširenje rada iz predmeta *Osnovi računarske inteligencije*, odbranjenog kod asistenta Aleksandra Vujinovića 2024/2025. školske godine. Metode rane i srednje integracije korišćene u prethodnom radu biće poređene zajedno sa estimacijom na osnovu jedne slike i metodom kasne integracije. Proširenje takođe uključuje implementaciju web platforme za određivanje lokacije upotrebom TensorFlow.js (baziran na WebGL-u), omogućavajući interaktivno poređenje sve četiri metode.

## Motivacija

Geokodovanje na osnovu slike može imati primenu u spasilačkim i državnim službama za pronalaženje osoba na osnovu fotografija, kao i u automatskom tagovanju lokacija slika za potrebe feed algoritama društvenih mreža.

## Skup podataka

Skup podataka je kreiran korišćenjem [Google Street View Static API](https://developers.google.com/maps/documentation/streetview/overview). Izabrane su zemlje unutar Evrope sa zadovoljavajućom pokrivenošću. Problem je predstavljen kao klasifikacija gde je jedna klasa jedna ćelija na mapi sa ivicom od 180 km. Dobijeno je 108 ćelija, za svaku je odabrano 300 lokacija, a za svaku lokaciju 5 slika iz različitih uglova (72° vidno polje, 640×460 piksela). Ukupno: 162.000 slika. Dataset: [link](https://ftp.moma.rs/VirgilNN/dataset.zip).

## Pretprocesiranje podataka

Rezolucija slika se prepolovljava na 320×230 piksela. Skup podataka se augmentuje vertikalnom refleksijom čime se udvostručava. Za neke metode, 5 slika se spajaju u jednu panoramu.

## Metodologija

Koristi se konvolucijska neuronska mreža (CNN) sa Inception ResNet arhitekturom istreniranom na [ImageNet](https://www.image-net.org/challenges/LSVRC/2012/) skupu (*transfer learning*). Porede se četiri pristupa:

- **Rana integracija**: 5 slika se spajaju u panoramu koja ulazi u model
- **Srednja integracija**: Svaka slika se zasebno obrađuje do feature mapa, koje se zatim objedinjuju pomoću *Deep Set* metode
- **Kasna integracija**: Svaka slika se nezavisno klasifikuje, rezultati se kombinuju max pooling-om
- **Jedna slika**: Klasifikacija korišćenjem samo jedne slike

## Evaluacija

Loss se računa kao *Cross-Entropy Loss*. Metode se porede korišćenjem *Top-N accuracy* metrike. Za predikciju koordinata koristi se ponderisana sredina koordinata centara klasa sa softmax vrednostima kao težinama. Greška se meri *Haversine* distancom.

**Comment by vdragan1993** [comment]
Pošto se na projektu u sklopu prethodnog predmeta već radilo sa Inception ResNet arhitekturom, za potrebe ovog projekta potrebno je predložiti 2 nove arhitekture (svaki član tima implementira jednu).
